### PR TITLE
[SPARK-20348] [ML] Support squared hinge loss (L2 loss) for LinearSVC

### PR DESCRIPTION
**Pull request by hhbyyh**
## What changes were proposed in this pull request?

While Hinge loss is the standard loss function for linear SVM, Squared hinge loss (a.k.a. L2 loss) is also popular in practice. L2-SVM is differentiable and imposes a bigger (quadratic vs. linear) loss for points which violate the margin. Some introduction can be found from http://mccormickml.com/2015/01/06/what-is-an-l2-svm/

Liblinear and scikit learn both offer squared hinge loss as the default loss function for linear SVM.

## How was this patch tested?

strengthen existing unit test and add new unit test for comparison.

### CODE REVIEW

File: mllib/src/main/scala/org/apache/spark/ml/classification/LinearSVC.scala [hunk 1]
```diff
@@ -42,15 +44,35 @@ import org.apache.spark.sql.functions.{col, lit}
...
+   * Specifies the loss function. Currently "hinge" and "squared_hinge" are supported.
+   * "hinge" is the standard SVM loss (a.k.a. L1 loss) while "squared_hinge" is the square of
+   * the hinge loss (a.k.a. L2 loss).
+   *
+   * @see <a href="https://en.wikipedia.org/wiki/Hinge_loss">Hinge loss (Wikipedia)</a>
+   *
+   * @group param
+   */
+  @Since("2.3.0")
+  final val lossFunction: Param[String] = new Param(this, "lossFunction", "Specifies the loss " +
```
Comment (yanboliang) [inline mllib/src/main/scala/org/apache/spark/ml/classification/LinearSVC.scala]: I'd prefer to move this out to shared params, since it can be used by other algorithms as well. Thanks.

Comment (hhbyyh) [inline mllib/src/main/scala/org/apache/spark/ml/classification/LinearSVC.scala]: Sure we can do it. 
But I'm thinking maybe we should conduct an integrated refactor about the common optimization parameters some time in the future, either through shared params or other trait or abstract class.

Comment (yanboliang) [inline mllib/src/main/scala/org/apache/spark/ml/classification/LinearSVC.scala]: OK, let leave as is and refactor in the future. One minor issue: What about renaming it to ```loss```? I found the name of corresponding params in ```sklearn.svm.linearSVC``` is ```loss```. Thanks.

**Comment by HyukjinKwon** [comment]
ping @hhbyyh, where are we on this?

**Comment by hhbyyh** [comment]
Hi @HyukjinKwon I think this is a feature we need, but currently we are still having some discussion about optimizer interface.

**Comment by HyukjinKwon** [comment]
I took out this in the list. Though, should we maybe close this for now and reopen again when it's ready if it takes quite long? It'd be probably better than leaving this open without further updates for long time.

**Comment by HyukjinKwon** [comment]
ping @hhbyyh. WDYT?

**Comment by hhbyyh** [comment]
OK. I'll close it for now and try to merge it with https://github.com/apache/spark/pull/17862.
Thanks for the comment from @yanboliang
